src/services/blocks/trace/Trace.ts

```diff
@@ -0,0 +1,666 @@
+import { ApiPromise } from '@polkadot/api';
+import { AccountInfo } from '@polkadot/types/interfaces';
+import { Registry } from '@polkadot/types/types';
+import { stringCamelCase } from '@polkadot/util';
+import BN from 'bn.js';
+import { InternalServerError } from 'http-errors';
+
+import { IOption, isSome } from '../../../types/util';
+import {
+	ActionGroup,
+	BlockTrace,
+	KeyInfo,
+	Operation,
+	PAccountEvent,
+	PActionEvent,
+	PalletKeyInfo,
+	PEvent,
+	Phase,
+	PhaseOther,
+	SpanWithChildren,
+	SpecialKeyInfo,
+	TraceSpan,
+} from './types';
+
+const EMPTY_KEY_INFO = {
+	error: 'key not found',
+};
+
+/**
+ * Known spans
+ */
+const SPANS = {
+	executeBlock: {
+		name: 'execute_block',
+		target: 'frame_executive',
+	},
+	applyExtrinsic: {
+		name: 'apply_extrinsic',
+		target: 'frame_executive',
+	},
+	onInitialize: {
+		name: 'on_initialize',
+		// Targets vary by pallet
+	},
+	onFinalize: {
+		name: 'on_finalize',
+		// Targets vary by pallet
+	},
+};
+
+// In the future this will need a more chain specific solution that creates
+// keys directly from expanded metadata.
+/**
+ * Get all the storage key names based on the ones built into `api`.
+ *
+ * @param api ApiPromise
+ */
+function getKeyNames(api: ApiPromise): Record<string, KeyInfo> {
+	const moduleKeys = Object.keys(api.query).reduce((acc, mod) => {
+		Object.keys(api.query[mod]).forEach((item) => {
+			const queryObj = api.query[mod][item];
+			let key;
+			try {
+				// Slice off '0x' prefix
+				key = queryObj.key().slice(2);
+			} catch {
+				key = queryObj.keyPrefix().slice(2);
```

Comment:
I didnt know this but if the `key` doesnt exist, there will always be a `keyPrefix`?

Reply:
https://github.com/paritytech/substrate-api-sidecar/issues/528

Reply:
I believe so. Reasoning is that anything under `query` is a storage item and all storage items live under some key or a key prefix if it something like a map where the non-prefix part of the storage key is the map key.

Reply:
[Here](https://github.com/polkadot-js/api/blob/a1f07e32743eed1c27bf06ca6fb9118afdbfa839/packages/api/src/base/Decorate.ts#L381-L384) you can see `key` expects either 1 or 2 args (depending on if map or double map). Where as keyPrefix simply expects an optional args. (`creator.keyPrefix` as a signature of `keyPrefix: (arg?: any) => Uint8Array;`)